### ISSUE TITLE
about dynamic OPENDRIVE

### ISSUE THREAD

**Issue by csawp** [esmini/esmini#334]
Whether to support dynamically adding a road or deleting a road
The initialization file of dynamic lane addition in midway is not XODR
OPENDRIVE similar to editing function

**Comment by eknabevcc** [comment]
Can you please elaborate on the issue, like what is the problem or what do you want to achieve?

Does it perhaps relate to a previously filed issue?

**Comment by csawp** [comment]
Many main roads and branch roads over 100km of expressway lead to slow loading speed, and need to be modified at any time between roads. Currently, to modify a road, it is necessary to release all ROAD

**Comment by eknabevcc** [comment]
Ok, I think I understand. 

Whenever you make a change in the road network you actually get a completely new OpenDRIVE file that needs to be loaded. And then esmini will create all OSI points (for lanes, boundaries and roadmarks) plus create the 3D geometry model of the road. 

I think the OSI point calculations is the bottleneck (most time consuming) and it's needed for the 3D model creation.

Unfortunately I don't see a quick fix for this. A potential approach would be to manipulate the road collection within esmini/roadmanager. For example, modifying or replacing just one road in the network, and then re-calculate OSI info and re-do the 3D model tessellation. 

But that would require some refactorization since currently the OSI calculations and 3D tessellations are done on the complete road network, and only once right after it (the OpenDRIVE) has been loaded. 

I can't support with actual implementation, but if you want to give it a shot I can try to support along the way and then consider a pull request of course.

**Comment by csawp** [comment]
My idea is to dynamically load each ROAD.

When the lane width is 0, there is no need to recalculate the OSI of the last lane.

Or the drawing of roads and markings does not depend on OSI. This will increase efficiency dozens of times.

Use multithreading to calculate the OSI of the running car, and set the ID parameter of the ROAD required for OSI

**Comment by eknabevcc** [comment]
I agree that when lane width is zero, there's no need to recalculate for the next-coming outer lane. However, I would expect this to be a rare special case, so I think the performance gain would be minimal in the typical/average case.

I agree that road and road markings not necessarily depend on OSI, since they are defined by the OpenDRIVE. And actually, back in the days esmini did create simple road feature visualization from the raw OpenDRIVE data. 

However, when we needed more detailed graphical representation of the the road network we concluded that the OSI points could be re-used for the road geometry tessellation. The OSI data includes all features like lane boundaries and lane markings. And more importantly it consider curvature, both X,Y and elevation, used for optimizing the tessellation (more triangles where needed).

Anyway, I think it would be feasible to do all this on road basis instead of complete road network. I mean, then we could have a loop over all roads that easily could be parallelized/multithreaded. At least it would be easy to just recalculate a single road whenever it has been modified. So even if OSI is recalculated it's only for a single road. 

The last comment I interpret as a suggestion to limit the OSI data to only the road or area close to the Ego. That's an interesting idea, but we haven't thought about it since we typically limit the scenario, including road network, to what's actually needed for the specific test case. 

Although interesting and good ideas, none of this will be prioritized from our side in near time. Let us know if you plan to implement some part of it, else I think we should close the issue for now.
